### PR TITLE
MSAL Runtime logging integration

### PR DESCRIPTION
**Pull request by SameerK-MSFT**
Fixes #
Adds unit tests

**Changes proposed in this request**
Added Unit tests

**Testing**
Unit Tests

**Performance impact**
None

**Documentation**
- [x] All relevant documentation is updated.
No doc needed

### CODE REVIEW

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 1]
```diff
@@ -83,6 +93,11 @@ private static void OnProcessExit(object sender, EventArgs e)
         {
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
+            if (_logger.IsDefaultPlatformLoggingEnabled)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:96]: Don't use `IsDefaultPlatformLoggingEnabled`. We should probably get rid of it, it's poorly documented. 

Use `_logger.IsLoggingEnabled()`

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 2]
```diff
@@ -27,6 +27,16 @@ internal class RuntimeBroker : IBroker
         private readonly WindowsBrokerOptions _wamOptions;
         private static Exception s_initException;
 
+        private readonly Dictionary<NativeInterop.LogLevel, LogLevel> _logLevelMap = new Dictionary<NativeInterop.LogLevel, LogLevel>()
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:30]: make it static

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:96]: CC @trwalke for help here.

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:96]: Quite possibly we should always assume a logger exists...

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 3]
```diff
@@ -94,6 +109,60 @@ private static void OnProcessExit(object sender, EventArgs e)
...
+        private void LogEventRaised(NativeInterop.Core sender, LogEventArgs args)
+        {
+            LogLevel msalLogLevel = _logLevelMap[args.LogLevel];
+            if (_logger.IsLoggingEnabled(msalLogLevel))
+            {
+                if (_logger.PiiLoggingEnabled)
+                {
+                    switch (msalLogLevel)
+                    {
+                        case LogLevel.Always:
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:121]: nit: you can get rid of the switches and just use 

`logger.Log(LogLevel, messageWithPii, messageWithoutPii)`

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 4]
```diff
@@ -153,7 +188,115 @@ public void HandleInstallUrl_Throws()
...
+        // Done - Logs in one
+        // NA - logs in multiple
+        // Done - maps correctly
+        // Done - Is Pii enabled
+        // Done - does not emit when loglevel is not appropriate
+        // Is API Platform logging enabled?
+        // no mem leaks
+
+        [TestMethod]
+        public async Task ATS_CallsLog_When_CalledAsync()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:201]: We have ROPC integration tests. Can you add real logging there and maybe assert that something comes from MSAL C++?

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 5]
```diff
@@ -83,6 +93,11 @@ private static void OnProcessExit(object sender, EventArgs e)
         {
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
+            if (_logger.IsDefaultPlatformLoggingEnabled)
+            {
+                s_lazyCore.Value.LogEvent += LogEventRaised;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:98]: This is an event handle leak. You keep adding handles, but you never remove them. 

I think you need to register the callback on every IBroker method and remove it in a "finally" block.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 2 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs]: Done.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 5 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs]: Done. In a different way.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 6]
```diff
@@ -94,10 +104,66 @@ private static void OnProcessExit(object sender, EventArgs e)
             }
         }
 
+        private void LogEventRaised(NativeInterop.Core sender, LogEventArgs args)
+        {
+            using LogEventWrapper logEventWrapper = new LogEventWrapper(this);
+            LogLevel msalLogLevel = LogLevelMap[args.LogLevel];
+            if (_logger.IsLoggingEnabled(msalLogLevel))
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:111]: In regards to [this comment](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3777#discussion_r1010386713)

This is the best way to do it. If the logger is not enabled, there should be a null logger which will always return null.

Comment (trwalke) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:111]: I meant to say false above. made a correction

Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs]: Done.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 7]
```diff
@@ -456,5 +528,38 @@ public bool IsBrokerInstalledAndInvokable(AuthorityType authorityType)
             _logger.Verbose($"[WAM Broker] MsalRuntime initialization successful.");
             return true;
         }
+
+        internal class LogEventWrapper : IDisposable
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:502]: This is used only by this class. So better to have it as part of  RuntimeBroker

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs (hunk 4 above)
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:200]: Added

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 1 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs]: Spoke with @trwalke .Please check the changes made. We can discuss if you disagree.

File: src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj [hunk 8]
```diff
@@ -4,6 +4,7 @@
     <TargetFrameworkNetStandard>netstandard2.0</TargetFrameworkNetStandard>
     <TargetFrameworkNetDesktop Condition="$([MSBuild]::IsOsPlatform('Windows')) and '$(MSAL_DESKTOP_ONLY_DEV)' == ''">net461</TargetFrameworkNetDesktop>
 
+    <LangVersion>8.0</LangVersion>
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj:7]: Inconsistent with MSAL? Why is this here?

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 9]
```diff
@@ -94,10 +104,36 @@ private static void OnProcessExit(object sender, EventArgs e)
...
+            if (_logger != null)
+            {
+                using LogEventWrapper logEventWrapper = new LogEventWrapper(this);
+                LogLevel msalLogLevel = LogLevelMap[args.LogLevel];
+                if (_logger.IsLoggingEnabled(msalLogLevel))
+                {
+                    string msgWithPii = string.Empty;
+                    string msgNoPii = string.Empty;
+
+                    if (_logger.PiiLoggingEnabled)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:118]: This is not how msal_runtime handles PII. They handle it by passing flag (in auth params?), in which case both exception emssages and log messages have PII.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 10]
```diff
@@ -435,26 +477,59 @@ public bool IsBrokerInstalledAndInvokable(AuthorityType authorityType)
...
+                _logger?.InfoPii(s_initException);
                 return false;
             }
 
-            _logger.Verbose($"[WAM Broker] MsalRuntime initialization successful.");
+            _logger?.Verbose($"[WAM Broker] MsalRuntime initialization successful.");
             return true;
         }
+
+        internal class LogEventWrapper : IDisposable
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:502]: Please move this to its own class.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 11]
```diff
@@ -94,10 +104,36 @@ private static void OnProcessExit(object sender, EventArgs e)
...
+                    string msgNoPii = string.Empty;
+
+                    if (_logger.PiiLoggingEnabled)
+                    {
+                        msgWithPii = args.Message;
+                    }
+                    else
+                    {
+                        msgNoPii = args.Message;
+                    }
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:125]: args.Message is the same thing. See my comment about PII.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
PII is not handled correctly.

File: src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj (hunk 8 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj:7]: 8.0 is required to declare using of a local variable. e.g.
``` CSharp
using LogEventWrapper logEventWrapper = new LogEventWrapper(this);
```

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/MsalLoggerValidator.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+using System;
+using Microsoft.IdentityModel.Abstractions;
+
+/// <summary>
+/// Performs validation on logs
+/// </summary>
+public class MsalLoggerValidator : IIdentityLogger
+{
+    // string that i tpart of MSAL logs
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/MsalLoggerValidator.cs:12]:
```suggestion
    // string that is part of MSAL logs
```

**Review by trwalke — APPROVED** [review]
Besides PII, LGTM

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 13]
```diff
@@ -153,7 +187,116 @@ public void HandleInstallUrl_Throws()
...
+            }
+            catch (MsalUiRequiredException)
+            {
+                // Assert
+                _logger.Received().Log(Arg.Any<Client.LogLevel>(), Arg.Any<string>(), Arg.Any<string>());
+            }
+            catch (Exception ex)
+            {
+                Assert.IsTrue(false, ex.Message);
+            }
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:236]: I think it's ok to remove this and for test to just fail by throwing.

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 14]
```diff
@@ -153,7 +187,116 @@ public void HandleInstallUrl_Throws()
...
+        [DataRow(Client.NativeInterop.LogLevel.Error, Client.LogLevel.Error)]
+        [DataRow(Client.NativeInterop.LogLevel.Fatal, Client.LogLevel.Error)]
+        public void LogEventRaised_MapsEvents_Correctly_When_No_Pii(Client.NativeInterop.LogLevel nativeLogLevel, Client.LogLevel msalLogLevel)
+        {
+            const string logMessage = "This is test";
+            _logger.IsLoggingEnabled(msalLogLevel).Returns(true);
+
+            Type wamBrokerType = _wamBroker.GetType();
+            MethodInfo fireLogMethod = wamBrokerType.GetMethod("LogEventRaised", BindingFlags.NonPublic | BindingFlags.Instance);
+            fireLogMethod.Invoke(_wamBroker, new object[] { null, new LogEventArgs(nativeLogLevel, logMessage) });
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:253]: Any way to test this without reflection? Make LogEventRaised internal?

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 15]
```diff
@@ -435,26 +477,59 @@ public bool IsBrokerInstalledAndInvokable(AuthorityType authorityType)
...
+
+        internal class LogEventWrapper : IDisposable
+        {
+            private bool _disposedValue;
+            RuntimeBroker _broker;
+
+            public LogEventWrapper(RuntimeBroker broker)
+            {
+                _broker = broker;
+                s_lazyCore.Value.LogEvent += _broker.LogEventRaised;
```
[the remark this answers is not in the document]
Comment (pmaytak) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:510]: I think it will be difficult to repro this manually. Maybe using blocking tasks, like when we test WithCacheSynchronization.

But it could happen if request 1 is disposing and unsubscribed the event handler, while request 2 is in the middle of doing work.

I forget, is he event collection a set or a list? Would calling `+= _broker.LogEventRaised;` multiple times, add multiple delegates or just 1?

**Review by pmaytak — CHANGES_REQUESTED** [review]

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 16]
```diff
@@ -81,7 +91,7 @@ private static void OnProcessExit(object sender, EventArgs e)
             ApplicationConfiguration appConfig,
             ILoggerAdapter logger)
         {
-            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:84]: @trwalke we discussed this please clarify.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 9 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:118]: Pii will not be supported in this release. Created a feature request. https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3822

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 11 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:125]: Updated as there will be no Pii in the first release

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/MsalLoggerValidator.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    private const string MSALCPPIdentifier = "[MSAL:000";
+
+    // count of MsalCPP logs
+    private int MsalCPPLogCount { get; set; }
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/MsalLoggerValidator.cs:16]: Good catch. I was using it earlier and was a leftover.

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 18]
```diff
@@ -1,40 +1,71 @@
...
-using Microsoft.Identity.Test.Common;
-using Microsoft.Identity.Client.Internal.Logger;
-using Microsoft.Identity.Client.Platforms.Features.DesktopOs.Kerberos;
-using System.Runtime.InteropServices;
-using System.Linq;
+using Account = Microsoft.Identity.Client.Account;
 
 namespace Microsoft.Identity.Test.Unit.BrokerTests
 {
+    internal class TestLogger : ILoggerAdapter
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:36]: Removed.

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 19]
```diff
@@ -153,7 +187,116 @@ public void HandleInstallUrl_Throws()
...
+        {
+            // Arrange
+            var appTokenCache = new TokenCache(_serviceBundle, isApplicationTokenCache: true);
+            var requestContext = new RequestContext(_serviceBundle, Guid.NewGuid());
+            var tenantAuthority = AuthorityInfo.FromAadAuthority(AzureCloudInstance.AzurePublic, tenant: TestConstants.AadTenantId, validateAuthority: false);
+            var acquireTokenCommonParameters = new AcquireTokenCommonParameters
+            {
+                ApiId = ApiEvent.ApiIds.AcquireTokenForClient,
+                AuthorityOverride = tenantAuthority
+            };
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:210]: This is done with OS current user. I am missing something. Let's discuss when you come back.

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs (hunk 14 above)
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:253]: I do not want to change the access modifier just of testing. If we can do testing with reflection, I avoid it.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:500]: > forget, is he event collection a set or a list? W
This is a collection and ad adn remove operations are thread safe Ref: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes

File: src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj (hunk 8 above)
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj:7]: I would like the latest. But we may face issues later as some of our users may not be able to use it. Still thinking 🤔

[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/Microsoft.Identity.Client.Broker.csproj:7]: Let's play safe and not use the latest to avoid any backward compatibility issues down the line.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:500]: Added parallel ATS for 25 ATS calls in the sample App (tests/devapps/WAM/NetFrameworkWam/Form1.cs). Still works.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 20]
```diff
@@ -94,10 +104,26 @@ private static void OnProcessExit(object sender, EventArgs e)
...
+        private void LogEventRaised(NativeInterop.Core sender, LogEventArgs args)
+        {
+            using LogEventWrapper logEventWrapper = new LogEventWrapper(this);
+            LogLevel msalLogLevel = LogLevelMap[args.LogLevel];
+            if (_logger.IsLoggingEnabled(msalLogLevel))
+            {
+                string msgWithPii = string.Empty;
+                string msgNoPii = string.Empty;
+
+                // as of now, there will be no Pii.
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:116]: Please add a link to the GH issue

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 20 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:116]: Done.

File: tests/Microsoft.Identity.Test.Unit/Microsoft.Identity.Test.Unit.csproj [hunk 21]
```diff
@@ -11,6 +11,8 @@
     <LangVersion>latest</LangVersion>
     
     <Configurations>Debug;Release;Debug + MobileApps</Configurations>
+
+    <Platforms>AnyCPU;x64</Platforms>
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/Microsoft.Identity.Test.Unit.csproj:15]: it was required for different platforms

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 22]
```diff
@@ -94,10 +104,27 @@ private static void OnProcessExit(object sender, EventArgs e)
             }
         }
 
+        private void LogEventRaised(NativeInterop.Core sender, LogEventArgs args)
+        {
+            using LogEventWrapper logEventWrapper = new LogEventWrapper(this);
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:109]: This should be removed since LogEventWrapper constructor adds LogEventRaised to the static event . So each time LogEventRaised is raised, we're adding one more handler.

**Review by pmaytak — APPROVED** [review]

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 22 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:109]: Good catch. That was copy/paste for each method added here too.

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 20 above)
Comment (pmaytak) [reply src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:115]: @SameerK-MSFT Was just thinking about this again. I may have misheard today in standup. So is this correct - errors from runtime time **do not** have PII; they are scrubbed? (I think I just heard you say the opposite).
